File: src/test/java/bad/robot/radiate/MonitoringTasksTest.java
package bad.robot.radiate;

import bad.robot.radiate.monitor.*;
import org.jmock.Expectations;
import org.jmock.integration.junit4.JUnitRuleMockery;
import org.junit.Rule;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ScheduledFuture;

import static bad.robot.radiate.RestartRequired.restartRequired;
import static java.util.Collections.emptyList;

public class MonitoringTasksTest {

    @Rule public final JUnitRuleMockery context = new JUnitRuleMockery();

    private final MonitoringTasksFactory factory = context.mock(MonitoringTasksFactory.class);
    private final Monitor monitor = context.mock(Monitor.class);
    
    @Test
    public void gathersTasks() {
        context.checking(new Expectations() {{
            ignoring(factory).notifyObservers(with(any(Exception.class)));
            oneOf(factory).create();
        }});
        new MonitoringTasks(factory, monitor);
    }

    @Test
    public void whenNoTasksAreGeneratedNotifyObserver() {
        context.checking(new Expectations() {{
            oneOf(factory).notifyObservers(with(any(NothingToMonitorException.class)));
            oneOf(factory).create(); will(returnValue(emptyList()));
        }});
        new MonitoringTasks(factory, monitor);
    }

    @Test
    public void exceptionsGatheringTasksNotifyObservers() {
        final Exception exception = new RuntimeException();
        context.checking(new Expectations() {{
            oneOf(factory).create(); will(throwException(exception));
            oneOf(factory).notifyObservers(exception);
            oneOf(factory).notifyObservers(restartRequired());
            oneOf(factory).notifyObservers(with(any(NothingToMonitorException.class)));
        }});
        new MonitoringTasks(factory, monitor);
    }

    @Test
    public void startAndStop() {
        final List<MonitoringTask> tasks = Arrays.<MonitoringTask>asList(new RandomStatus());
        final Iterable<ScheduledFuture<?>> scheduled = emptyList();
        context.checking(new Expectations() {{
            oneOf(factory).create(); will(returnValue(tasks));
            oneOf(monitor).start(tasks); will(returnValue(scheduled));
            oneOf(monitor).stop(scheduled);
        }});

        MonitoringTasks monitoring = new MonitoringTasks(factory, monitor);
        monitoring.start();
        monitoring.stop();
    }

}

File: src/test/scala/bad/robot/radiate/MonitoringTasksTest.scala
package bad.robot.radiate

import java.util.concurrent.ScheduledFuture

import bad.robot.radiate.monitor._
import bad.robot.radiate.scalamock._
import bad.robot.radiate.specs2.monitoringTasksAsResult
import org.scalamock.specs2.IsolatedMockFactory
import org.specs2.mutable.Specification

import scalaz.-\/
import scalaz.syntax.either._

class MonitoringTasksTest extends Specification with IsolatedMockFactory {

  val factory = mock[MonitoringTasksFactory]
  val monitor = mock[Monitor]

  "Gathers tasks" >> {
    val task = stub[MonitoringTask]

    (factory.notifyObservers(_: Exception)).expects(*).anyNumberOfTimes
    (factory.create _).expects().once.returning(List(task).right)

    new MonitoringTasks(factory, monitor)
  }

  "When no tasks are generated, notify observer" >> {
    (factory.notifyObservers(_: Exception)).expects(anyTypedOf[NothingToMonitorException]).once
    (factory.create _).expects().once.returning(List().right)

    new MonitoringTasks(factory, monitor)
  }

  "Exceptions thrown when gathering tasks will notify observers" >> {
    val arbitaryError = ParseError("example")

    (factory.notifyObservers(_: Error)).expects(arbitaryError).once
    (factory.notifyObservers(_: Information)).expects(RestartRequired()).once
    (factory.notifyObservers(_: Exception)).expects(anyTypedOf[NothingToMonitorException]).once
    (factory.create _).expects().once.returning(-\/(arbitaryError))

    new MonitoringTasks(factory, monitor)
  }

  "Start and stop" >> {
    val tasks = List(new RandomStatus)
    val scheduled = List.empty[ScheduledFuture[_]]

    (factory.create _).expects().returning(tasks.right).once
    (monitor.start _).expects(tasks).returning(scheduled).once
    (monitor.cancel _).expects(scheduled).once

    val monitoring = new MonitoringTasks(factory, monitor)
    monitoring.start()
    monitoring.stop()

    true must_== true
  }
}
